### PR TITLE
Improvements to fetch APIs

### PR DESCRIPTION
**Pull request by mislav**
In order of importance (more info in individual commit messages):

- Allow **any** string as HTTP method, not just `GET` | `POST` | `DELETE` |
  `HEAD` | `OPTIONS` | `PUT` | `PATCH` | `TRACE`

- Allow URL instance as 1st argument to `fetch()`, `new Request()`

- Add `Headers#forEach()`, `URLSearchParams#forEach()`

- Add `Reponse#redirected`, `Response#trailer`

- Mark the 2nd argument to `Response.redirect()` as optional

- Bring `ReferrerPolicyType` up to date

- Add `window` property to RequestOptions

Backwards incompatible changes:

- No more `MethodType` enum
- Remove `Headers#getAll()`
- These `RequestOptions` parameters are not nullable anymore:
  - `cache`
  - `credentials`
  - `headers`
  - `integrity`
  - `method`
  - `mode`
  - `redirect`
  - `referrer`
  - `referrerPolicy`
- These `ResponseOptions` parameters are not nullable anymore:
  - `status`
  - `statusText`
  - `headers`
- Remove `Response#useFinalURL`

Ref. https://github.com/facebook/flow/pull/1386

/cc @runn1ng @samwgoldman @AprilArcus

### CODE REVIEW

**Comment by runn1ng** [comment]
Nice!

I am never sure what the difference is between `cache?: ?CacheType` and `cache?: CacheType` (in RequestOptions) and what are the consequences of the change. :)

**Comment by mislav** [comment]
@runn1ng `cache?: ?CacheType` means that, in addition to `cache` property being optional in the options hash, its value can either be `CacheType` or null. However, since it doesn't make sense that you ever pass `cache: null` or `method: null` to a `fetch()` or `new Request()` invocation, the change that I've made here causes a type violation to be raised in those cases.

File: lib/bom.js [hunk 1]
```diff
@@ -812,8 +812,8 @@ declare class Headers {
     append(name: string, value: string): void;
     delete(name: string): void;
     entries(): Iterator<[string, string]>;
+    forEach((value: string, name: string, index: number) => any): void;
```
Comment (dgraham) [inline lib/bom.js]: The third callback argument is the Headers instance.

```js
new Headers({a:10, b:20}).forEach((a,b,c,d) => console.log(a,b,c,d))
// => 20 b Headers {  } undefined 
// => 10 a Headers {  } undefined
```

https://heycam.github.io/webidl/#es-forEach

Comment (mislav) [inline lib/bom.js]: Good catch; thanks! I've also noticed that I'm missing the `thisArg` optional argument.

**Comment by mislav** [comment]
I have corrected arguments to `forEach`.

**Comment by mislav** [comment]
@samwgoldman I just found another small tweak I want to do to the Response class. Now that this PR has been imported, can I keep updating this branch, or should I wait for this to be merged and open a new PR?

**Comment by samwgoldman** [comment]
Update away! I can sync again

**Comment by mislav** [comment]
@samwgoldman I've added one extra commit. Thank you!

**Comment by samwgoldman** [comment]
Thanks!

**Pull request closed without merging**